### PR TITLE
Add the ability to lock the waveform display to a specific time location

### PR DESCRIPTION
**Pull request by peppy**
Quite often you will want to be timing using a specific section of the waveform you've identified to have a discernible beat. This allows locking the display to that region while you make further adjustments.

Had to restructure a bit to allow displaying around a time offset, rather than a beat index. This is because the beat index will be changing along with the BPM.

https://user-images.githubusercontent.com/191335/171145067-136e9d0d-7460-4017-a48a-dcc5149c16f2.mp4

This also adds animation to adjustments coming from an external source. The reasoning for this will become more apparent after.. adjustment controls exist. But you can test using the test scene for now.

### CODE REVIEW

File: osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs [hunk 1]
```diff
@@ -156,23 +199,83 @@ private void showFrom(int beatIndex)
...
+            }
+
+            protected override void LoadComplete()
+            {
+                base.LoadComplete();
+
+                text
+                    .FadeIn().Then().Delay(500)
+                    .FadeOut().Then().Delay(500)
+                    .Loop();
```
Comment (bdach) [inline osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs:277]: Could this... not blink? Or at least do so slower? It felt really distracting to see this blinking red box when I saw this the first time, and it's blinking so fast that I can barely read the "locked" text in time before the text disappears between blinks.

File: osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs [hunk 2]
```diff
@@ -156,23 +199,83 @@ private void showFrom(int beatIndex)
...
+                double time = selectedGroupStartTime + index * timingPoint.BeatLength;
 
                 float offset = (float)(time - visible_width / 2) / trackLength * scale;
 
                 row.Alpha = time < selectedGroupStartTime || time > selectedGroupEndTime ? 0.2f : 1;
-                row.WaveformOffset = -offset;
+                row.WaveformOffsetTo(-offset, animated);
                 row.WaveformScale = new Vector2(scale, 1);
-                row.BeatIndex = beatIndex++;
+                row.BeatIndex = (int)Math.Floor(index);
```
Comment (bdach) [inline osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs:217]: It confused me for a good while why `index` is permitted to be fractional all the way up to this point (thinking that the display was kind of "lying" by pretending that it's showing waveforms for beats, while it is actually showing parts of neighbouring bars glued together). After trying to floor the index more aggressively I get why that is the case now (leads to smoother display when locked, and the waveform will resnap to a correct beat on next `Update()` after unlocking), but maybe worth adding an inline comment?

Comment (peppy) [reply osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs:217]: It has to be fractional to support display at arbitrary time. Else that pathway would not work correctly.

I considered rewriting this method to not use beat index but it didn't seem worthwhile.

Comment (bdach) [reply osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs:217]: I get why it has to be fractional, it's just more that it felt like a bit of a lie to be showing integral beat indices while the current time instant is not snapped to a beat. I'd honestly consider hiding the indices in the mean time maybe?

File: osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs [hunk 3]
```diff
@@ -81,72 +88,108 @@ protected override void LoadComplete()
...
-            selectedGroupEndTime = beatmap.Value.Track.Length;
-
             var tcp = selectedGroup.Value?.ControlPoints.OfType<TimingControlPoint>().FirstOrDefault();
 
             if (tcp == null)
             {
                 timingPoint = new TimingControlPoint();
+                // During movement of a control point's offset, this clause can be hit momentarily.
+                // We don't want to reset the `selectedGroupStartTime` here as we rely on having the
+                // last value to perform an offset traversal below.
```
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs:117]: If I were to suggest something:

```suggestion
                // During movement of a control point's offset, this clause can be hit momentarily,
                // as moving a control point is implemented by removing it and inserting it at the new time.
                // We don't want to reset the `selectedGroupStartTime` here as we rely on having the
                // last value to update the waveform display below.
```

File: osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs (hunk 1 above)
Comment (bdach) [reply osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs:277]: Doubling the durations to a second seems better to me, but then again I really wouldn't mind just having the text stay there for the entirety of the display lock.

Comment (peppy) [reply osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs]: I've increased the delay to 600ms (the longest I can make it without it feeling bad) and also ensured the text shows immediately on initial show.

**Review by bdach — APPROVED** [review]
let's see how this holds up in user testing I guess

**Pull request merged**